lib/firebase/admin/fcm/message_encoder.rb

```diff
@@ -0,0 +1,355 @@
+          elsif result.key?(:"title-loc-args") && !result.key?(:"title-loc-key")
+            raise ArgumentError, "APSAlert.title_loc_key is required when specifying title_loc_args"
+          elsif result.key?(:"subtitle-loc-args") && !result.key?(:"subtitle-loc-key")
+            raise ArgumentError, "APSAlert.subtitle_loc_key is required when specifying subtitle_loc_args"
+          end
+
+          if (custom_data = alert.custom_data)
+            raise ArgumentError, "APSAlert.custom_data must be a hash" unless custom_data.is_a?(Hash)
+            custom_data.each do |k, v|
+              unless k.is_a?(String) || k.is_a?(Symbol)
+                raise ArgumentError, "APSAlert.custom_data key #{k}, must be a string or symbol"
+              end
+              k = k.to_sym
+              result[k] = v
+            end
+          end
+          remove_nil_values(result)
+        end
+
+        # @return [Hash, String, nil]
+        def encode_aps_sound(sound)
+          return nil unless sound
+          return sound if sound.is_a?(String) && !sound.empty?
+          unless sound.is_a?(CriticalSound)
+            raise ArgumentError, "APS.sound must be a non-empty string or a CriticalSound"
+          end
+
+          result = {
+            name: check_string("CriticalSound.name", sound.name, non_empty: true),
+            volume: check_numeric("CriticalSound.volume", sound.volume)
+          }
+
+          result[:critical] = 1 if sound.critical
+          raise ArgumentError, "CriticalSound.name is required" if result[:name].nil?
+
+          if (volume = result[:volume])
+            raise ArgumentError, "CriticalSound.volume must be between [0,1]." unless volume >= 0 && volume <= 1
+          end
+
+          remove_nil_values(result)
+        end
+
+        # @return [Hash, nil]
+        def encode_notification(v)
+          return nil unless v
+          raise ArgumentError, "Message.notification must be a Notification" unless v.is_a?(Notification)
+          result = {
+            body: check_string("Notification.body", v.body),
+            title: check_string("Notification.title", v.title),
+            image: check_string("Notification.image", v.image)
+          }
+          remove_nil_values(result)
+        end
+
+        # @return [Hash, nil]
+        def encode_fcm_options(options)
+          return nil unless options
+          raise ArgumentError, "Message.fcm_options must be a FCMOptions." unless options.is_a?(FCMOptions)
+          result = {
+            analytics_label: check_analytics_label("Message.fcm_options", options.analytics_label)
+          }
+          remove_nil_values(result)
+        end
+
+        # Remove nil values and empty collections from the specified hash.
+        # @return [Hash]
+        def remove_nil_values(hash)
+          hash.reject do |_, v|
+            if v.is_a?(Hash) || v.is_a?(Array)
+              v.empty?
+            else
+              v.nil?
+            end
+          end
```

Comment:
It seems to me that since you return `nil` in the `check_` functions for absent values, you can replace `remove_nil_values` with just `hash.compact`